### ISSUE TITLE
Collect template paths during request

### ISSUE THREAD

**Issue by martinRenou** [QuantStack/voila#258]
cc. @maartenbreddels @SylvainCorlay @jtpio 

From what I understand of the code base, the way `template_paths` are collected feels bad to me. The paths are collected once and computed given the template name, when starting the `app` or the `server_extension`. Then the request handler always uses the same list of template paths, which makes it quite impossible to correctly choose the right template (e.g. when trying to choose the template from the `url` like @jtpio is doing in this PR #201).

I think the proper way of doing it would be to do a `collect_template_paths` for each request.

**Comment by jtpio** [comment]
https://github.com/QuantStack/voila/pull/201 is collecting the template paths based on the folder structure. Then all templates can be found without specifying a template name anymore.

But it is still done only once at startup that's right.

**Comment by maartenbreddels** [comment]
That is exactly why I want to change the template system, good you see it the same way. Once https://github.com/QuantStack/voila/pull/208 is in, this will be much cleaner.

**Comment by martinRenou** [comment]
> #201 is collecting the template paths based on the folder structure. Then all the templates can be found without specifying a template name anymore.

Oh I see, then we collect everything. The thing is having to sort the list of template paths later feels bad to me... There should be a cleaner way of doing it.

> That is exactly why I want to change the template system, good you see it the same way. Once #208 is in, this will be much cleaner.

It is still not clear in my head what you are doing in your PR, I'll take a better look at it.

Maybe we should all discuss together, there will be conflicts with those two PRs

**Comment by maartenbreddels** [comment]
Yes, as I said, it would be good to discuss this over a high bandwidth channel (video). It is good you've hit the same issues, that makes it easier to understand. Let's try on Thursday!

**Comment by jtpio** [comment]
Yes that would be great.

In the end #201 and #208 complement each other:
- #201 is about going through all the installed templates and making them available to the end user
- #208 is about refactoring how these templates are stored on disk and potentially changing the search path

**Comment by martinRenou** [comment]
> #201 is about going through all the installed templates and making them available to the end user

I would rather keep collecting the template paths for a specific template name, but doing it in the handler (per request). What do you think? Could it be slow?

**Comment by maartenbreddels** [comment]
I thought about that as well, it might be slow indeed, I think we want to cache that.

**Comment by martinRenou** [comment]
Fair enough, so another idea: instead of having a list of template paths, maybe a dictionary 
```
{ "template_name": [template_paths, ...], ... }
``` 
?

**Comment by jtpio** [comment]
> I would rather keep collecting the template paths for a specific template name, but doing it in the handler (per request). What do you think? Could it be slow?

Also the installed templates are not likely to change that often, normally only after pip installing one. Caching makes sense.

**Comment by timkpaine** [comment]
> Also the installed templates are not likely to change that often, normally only after pip installing one. Caching makes sense.

Not if your users are developing a custom template for their application on your voila-based platform. We dynamically dump templates to disk on startup.

**Comment by maartenbreddels** [comment]
I think that assuming the directory structure exists is a fair assumption right? The files/content can freely change though (except maybe the template name it inherits from). If we don't cache templatename->directory paths mapping, requests may be quite expensive.

**Comment by martinRenou** [comment]
So, if we want to cache a `template_name` -> `template_paths` mapping, my question is: Can we easily retrieve all the possible values for `template_name` when starting the app/server extension?
